### PR TITLE
[Play] 2048 Game

### LINKED ISSUE #577: [Add a Play]: 2048 Game
## What's the Play about?
I want to build a clone of the 2048 game in React with animations and some other features.

## What ReactJS concept will be used to create this play?
JSX, useState, useEffect, hooks, and Props

## Are you willing to take it up for implementation?
Yes!!

### PR DESCRIPTION
**Pull request by Shivam-Katare**
# Description

This play is the clone of the game 2048 game. This game contains the feature of the best score, animations, and some sound effects.

Fixes #577 

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by atapas** [comment]
@Shivam-Katare is it review ready?

**Comment by Shivam-Katare** [comment]
> @Shivam-Katare is it review ready?

Yes @atapas, it is ready for review.

**Review by ammaaraslam — CHANGES_REQUESTED** [review]
It's an awesome play @Shivam-Katare! Can't resist playing it.
Just a few suggestions
- After the "Game Over" popup is displayed, when you click on the help button, the help popup seems to overlap with the game over popup. To fix this you could either not allow the user to click on the help button after the game over popup. Or trigger a function to close the game over popup when the help button is clicked.

image

- Try make the content a bit bigger and occupy more space

Everything else looks great!

File: src/plays/2048/Game2048Components/Cell.js
Comment (Sachin-chaurasiya) [inline src/plays/2048/Game2048Components/Cell.js:5]:
```suggestion
    <div className={`cell cell--${props.label}`}>
```

File: src/plays/2048/Game2048Components/Game.js
Comment (Sachin-chaurasiya) [inline src/plays/2048/Game2048Components/Game.js:17]: Extract this to function and call that function here.

File: src/plays/2048/Game2048Components/Game.js
Comment (Sachin-chaurasiya) [inline src/plays/2048/Game2048Components/Game.js:21]:
```suggestion
    <div className="game_wrapper">
```

File: src/plays/2048/Game2048Components/Game.js
Comment (Sachin-chaurasiya) [inline src/plays/2048/Game2048Components/Game.js:24]: @Shivam-Katare, Do not pass the setter function as a prop, instead create a handler and pass that function as a prop.

Example : 
```js
const handleGrid=(value)=>setGrid(value)
```

File: src/plays/2048/Game2048Components/GameOverlay.js
Comment (Sachin-chaurasiya) [inline src/plays/2048/Game2048Components/GameOverlay.js:14]:
```suggestion
```

File: src/plays/2048/Game2048Components/GameOverlay.js
Comment (Sachin-chaurasiya) [inline src/plays/2048/Game2048Components/GameOverlay.js:47]:
```suggestion
```

File: src/plays/2048/Game2048Components/GameOverlay.js
Comment (Sachin-chaurasiya) [inline src/plays/2048/Game2048Components/GameOverlay.js:12]: @Shivam-Katare , Can we add some comments here?

File: src/plays/2048/Game2048Components/GameOverlay.js
Comment (Sachin-chaurasiya) [inline src/plays/2048/Game2048Components/GameOverlay.js:17]: @Shivam-Katare , Can we take this check in a variable and use it in all the places.

Example:
```js
const isStatusOver= props.status === "over"

const isStatusWon= props.status === "won"

```

File: src/plays/2048/Game2048Components/Grid.js
Comment (Sachin-chaurasiya) [inline src/plays/2048/Game2048Components/Grid.js:32]: @Shivam-Katare, Extract `handleKeyDown` logic in a function and call that function in the useEffect.

File: src/plays/2048/Game2048Components/Modal.js
Comment (Sachin-chaurasiya) [inline src/plays/2048/Game2048Components/Modal.js:9]:
```suggestion
    setModal(previous=>!previous);
```

File: src/plays/2048/Game2048Components/Modal.js
Comment (Sachin-chaurasiya) [inline src/plays/2048/Game2048Components/Modal.js:16]: @Shivam-Katare , Can we add this logic in a useEffect and add the `modal` as a dependency, so whenever the modal state changes this logic gets executed?

File: src/plays/2048/Game2048Components/VideoMusic.js
Comment (Sachin-chaurasiya) [inline src/plays/2048/Game2048Components/VideoMusic.js:2]:
```suggestion
import Video from "../GameAudio/VideoMusic.mp4"
```

File: src/plays/2048/GameTwentyFourtyEight.js
Comment (Sachin-chaurasiya) [inline src/plays/2048/GameTwentyFourtyEight.js:34]: @Shivam-Katare , Same here create a handler for `setStatus` and pass that as a prop.

File: src/plays/2048/gameLogic.js
Comment (Sachin-chaurasiya) [inline src/plays/2048/gameLogic.js:15]:
```suggestion
  const score = props.score + calcScore(props.grid, grid);
```

File: src/plays/2048/gameLogic.js
Comment (Sachin-chaurasiya) [inline src/plays/2048/gameLogic.js:25]:
```suggestion
    score,
```

**Review by Sachin-chaurasiya — DISMISSED** [review]

**Comment by atapas** [comment]
@Shivam-Katare Can we wrap this up?

**Comment by Shivam-Katare** [comment]
> @Shivam-Katare Can we wrap this up?

Hey @atapas, I am working on the issues that **Sachin** pointed out, and I will complete them by tomorrow. After that, you can wrap this up.

**Review by ammaaraslam — COMMENTED** [review]
@Shivam-Katare 
Everything looks good...
Any plan to add a button to play/pause music?

**Comment by Shivam-Katare** [comment]
> @Shivam-Katare Everything looks good... Any plan to add a button to play/pause music?

@ammaaraslam Yes I tried to add but it was giving some conflicts. Like if the user doesn't pause the music, the music will continue after exiting the play. That's why I removed that button for now.

**Comment by ammaaraslam** [comment]
> > @Shivam-Katare Everything looks good... Any plan to add a button to play/pause music?
> 
> @ammaaraslam Yes I tried to add but it was giving some conflicts. Like if the user doesn't pause the music, the music will continue after exiting the play. That's why I removed that button for now.

I see.... no problem everything else is good now

**Review by ammaaraslam — APPROVED** [review]

File: src/plays/2048/Game2048Styles/Modal.css
Comment (atapas) [inline src/plays/2048/Game2048Styles/Modal.css:1]: @Shivam-Katare Please consider adding unique style names in this file too.

**Review by atapas — CHANGES_REQUESTED** [review]

File: src/plays/2048/Game2048Styles/Modal.css
Comment (Shivam-Katare) [reply src/plays/2048/Game2048Styles/Modal.css:1]: Okay.

**Comment by Shivam-Katare** [comment]
Hey, @atapas I have resolved all the issues you can review it now.

**Review by ammaaraslam — CHANGES_REQUESTED** [review]
@Shivam-Katare
Instead of using `camelCase`, the best practice for css classNames is to use `kebab-case`.

Like for the className `twentyFourtyEight`, use something like `twenty-forty-eight` or `tfe` for short

**Comment by Shivam-Katare** [comment]
> twenty-forty-eight

Hey @ammaaraslam, do I change this everywhere?

**Comment by ammaaraslam** [comment]
> > twenty-forty-eight
> 
> Hey @ammaaraslam, do I change this everywhere?

Yup that'd be good :)

**Comment by Shivam-Katare** [comment]
> > > twenty-forty-eight
> > 
> > 
> > Hey @ammaaraslam, do I change this everywhere?
> 
> Yup that'd be good :)

Okay

**Comment by Shivam-Katare** [comment]
> @Shivam-Katare Instead of using `camelCase`, the best practice for css classNames is to use `kebab-case`.
> 
> Like for the className `twentyFourtyEight`, use something like `twenty-forty-eight` or `tfe` for short

Hey @ammaaraslam, I changed the `camelCase` classNames to `kebab-case`. You can review it now.

**Review by atapas — APPROVED** [review]
Looks Good to me

**Comment by atapas** [comment]
@ammaaraslam Waiting for your final approval before merge. Please test it on the latest preview build before approving.

**Review by ammaaraslam — APPROVED** [review]
LGTM

**Pull request merged**
